### ISSUE TITLE
ZIO.Chunk classCast exception

### ISSUE THREAD

**Issue by ahoy-jon** [zio/zio#3242]
I am working on  #3241 

I created this test : 
```
    test("heterogeneous") {
      val h: Chunk.NonEmpty[AnyVal] = Chunk(1, 1.0, true)
      val m: Chunk.NonEmpty[AnyVal] = h.map(identity)

      assert(h)(equalTo(m))
    }
```

with this error 
```
java.lang.ClassCastException: class java.lang.Double cannot be cast to class java.lang.Integer (java.lang.Double and java.lang.Integer are in module java.base of loader 'bootstrap')
```

Is that behavior normal?
Should we update the documentation in https://github.com/zio/zio/blob/master/docs/datatypes/chunk.md ?

**Comment by adamgfraser** [comment]
Does this replicate if you do with normal `Chunk` constructor on `master`?

**Comment by ahoy-jon** [comment]
Very good question! 😃 It is the same on "1.0.0-RC18-2+73-c6e7b4e6-SNAPSHOT"


The bug is "simple", the classTag is obtained on the first element. Here I am using an Int, to start the chunk, it is expected to follow with ints.

If we try to create a chunk with :
```
Chunk(1, "a")
```
It doesn't work again.

This is a bad practice to create a `Chunk[AnyVal]` or a `Chunk[Any]`, but it is not documented.

That being said, the compiler is already complaining in warnings, like : 
*  a type was inferred to be `AnyVal`; this may indicate a programming error.
*  a type was inferred to be `Any`; this may indicate a programming error.


```
scala> import zio._
import zio._

scala> Chunk(1, 1.0, true)
<console>:15: warning: a type was inferred to be `AnyVal`; this may indicate a programming error.
       Chunk(1, 1.0, true)
             ^
java.lang.ClassCastException: class java.lang.Double cannot be cast to class java.lang.Integer (java.lang.Double and java.lang.Integer are in module java.base of loader 'bootstrap')
  at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:101)
  at scala.runtime.ScalaRunTime$.array_update(ScalaRunTime.scala:92)
  at scala.collection.IndexedSeqOptimized$class.copyToArray(IndexedSeqOptimized.scala:180)
  at scala.collection.mutable.WrappedArray.copyToArray(WrappedArray.scala:35)
  at scala.collection.TraversableOnce$class.copyToArray(TraversableOnce.scala:278)
  at scala.collection.AbstractTraversable.copyToArray(Traversable.scala:104)
  at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:286)
  at scala.collection.mutable.WrappedArray.toArray(WrappedArray.scala:73)
  at zio.Chunk$.fromIterable(Chunk.scala:893)
  at zio.Chunk$.apply(Chunk.scala:795)
  ... 40 elided

scala> 
```

**Comment by iravid** [comment]
This is by design. Would be great to add a warning to Chunk about this.

**Comment by adamgfraser** [comment]
Closed by #3249.
